### ISSUE TITLE
ModuleNotFoundError: Module not found: Error: Can't resolve 'dgram' in [...]

### ISSUE THREAD

**Issue by schlenger** [adzialocha/osc-js#50]
`dgram` seems to be deprecated/hold back: https://www.npmjs.com/package/dgram

Currently, there is no compilation possible with the setup.

**Comment by tschnz** [comment]
This happens when using node.js with webpack due to breaking changes upgrading webpack from v4 to v5. In short: When it comes to polyfills, v4 automatically sets all requires of node.js internals (fs, path, dgram, ...) to false packing for the browser. In v5 we explicitly have to state that there is NO polyfill, otherwise it triggers a warning, leaving osc-js useless.

**Solutions**

- Use webpack < v5
- Use a [fallback](https://webpack.js.org/configuration/resolve/#resolvefallback):
    - In webpack.config.js:
       ```
      module.exports = {
         //...
         resolve: {
           fallback: {
            "dgram": false, // do not include a polyfill for dgram,
           },
         },
       };
      ```
    - or when using the API in a build script:
      ```
       const webpackCompiler = webpack({
         // ...
         resolve: {
           fallback: { "dgram": false }
         },
      ```

**Comment by felixroos** [comment]
This also happens with webpack in the browser, as osc-js thinks its node when __dirname is set, see https://github.com/adzialocha/osc-js/blob/d5693c6cc4013fe39658a233ca37f44ff3e4b03a/src/plugin/dgram.js#L1 . 

The Problem: webpack will set __dirname in the browser too, which results in webpack trying to resolve dgram in the browser, which fails. 

Yes, the problem can be avoided with a custom webpack config, but thats not a perfect solution, because sometimes, you will not have direct access to the webpack config. For example in create-react-app (on of the most widely used ways to create frontend projects), you will only get to configure webpack after ejecting, which you would want to avoid. 

An easy fix would be to use window or global to differentiate between node and browser, see https://stackoverflow.com/questions/4224606/how-to-check-whether-a-script-is-running-under-node-js

**Comment by adzialocha** [comment]
> An easy fix would be to use window or global to differentiate between node and browser,

Thank you for this hint @felixroos, I'll look into it now! :smiley:

**Comment by adzialocha** [comment]
I've published a fix in release `v2.3.1`, could you tell me if this works for you @felixroos and @schlenger?

**Comment by yaxu** [comment]
That didn't work for me unfortunately. It works if a try-catch is added, like:
```javascript
  var dgram;
  try {
    dgram = require('dgram');
  } catch(error) {
    if (typeof window === 'undefined') {
      throw(error);
    }
  }
```
... and the same for `dgram$1`.

However, it doesn't seem to be working for @felixroos even then..

**Comment by adzialocha** [comment]
> That didn't work for me unfortunately. It works if a try-catch is added

Alright, I've made some experiments and think the best solution is to tackle the problem during the build process by:

1. Introducing a `browser` field in `package.json` which points at the special browser build
2. Replace `dgram` and `ws` with custom files during browser builds to have better control than just checking global contexts

It works for me with Webpack / simple static html websites and in Node :+1:

See PR here: https://github.com/adzialocha/osc-js/pull/63 I'll do some more testing and merge, release this soon!

**Comment by adzialocha** [comment]
This should be fixed now in `v2.3.2`. Let me know if the problem still persists, then I'll re-open this issue :+1:

**Comment by yaxu** [comment]
Brilliant, thanks!
